compiler/src/main/java/dagger/internal/codegen/BindingGraph.java

```diff
@@ -210,14 +215,26 @@ private BindingGraph create(
         explicitBindingsBuilder.add(provisionBindingFactory.forComponent(componentDependency));
         List<ExecutableElement> dependencyMethods =
             ElementFilter.methodsIn(elements.getAllMembers(componentDependency));
+        HashMultimap<Key, Equivalence.Wrapper<TypeMirror>> componentMethodType = HashMultimap.create();
```

Comment:
FWIW, I think it makes sense to solve this in [ComponentDescriptor](https://github.com/google/dagger/blob/e9066f673ed62cc736ca9c2671efbc3e1e9c2a49/compiler/src/main/java/dagger/internal/codegen/ComponentDescriptor.java#L410-L418), and then here to loop over componentDescriptor.dependencMethodIndex().keySet() so that the work is all being done twice.

The core of this bug appears to be weird, and I think it stems from Java8 compatibility (but I'm not positive).

In Java 8, the following is not allowed:

```java
interface A {
  default Integer number() {
    return 1;
  }
}

interface B {
  default Integer number() {
    return 2;
  }
}

interface C extends A, B {}
```

`C` needs to define which implementation of `number()` to use.

Anyway, getting back here, if the method signature is the same, but the two interfaces have covariant return types, Dagger will correctly generate bindings for both return types. E.g.:

```java
interface A {}
interface A1 extends A {}
interface A2 extends A1 {}

interface Base {
  A a();
}

interface ParentA extends Base {
  @Override A1 a();
}

interface ParentB extends Base {
  @Override A2 a();
}

interface Combined extends ParentA, ParentB {}

@Component(dependencies = Combined.class)
interface C {
  A1 a1();
  A2 a2();
}
```

So we basically just want to handle the case when the methods have no args, and the same name and return type. We do something similar (but not identical) in [AbstractComponentWriter](https://github.com/google/dagger/blob/e9066f673ed62cc736ca9c2671efbc3e1e9c2a49/compiler/src/main/java/dagger/internal/codegen/AbstractComponentWriter.java#L667-L669)